lib/util/template/index.ts

```diff
@@ -162,28 +161,35 @@ const allowedFieldsList = Object.keys(allowedFields)
 
 type CompileInput = Record<string, unknown>;
 
-type FilteredObject = Record<string, CompileInput | CompileInput[] | unknown>;
-
-function getFilteredObject(input: CompileInput): FilteredObject {
-  const obj = clone(input);
-  const res: FilteredObject = {};
-  const allAllowed = [
-    ...Object.keys(allowedFields),
-    ...exposedConfigOptions,
-  ].sort();
-  for (const field of allAllowed) {
-    const value = obj[field];
+const allowedTemplateFields = new Set([
+  ...Object.keys(allowedFields),
+  ...exposedConfigOptions,
+]);
+
+const compileInputProxyHandler: ProxyHandler<CompileInput> = {
+  get(target: CompileInput, prop: keyof CompileInput): unknown {
+    if (!allowedTemplateFields.has(prop)) {
+      return undefined;
+    }
+
+    const value = target[prop];
+
     if (is.array(value)) {
-      res[field] = value
+      return value
         .filter(is.plainObject)
-        .map((element) => getFilteredObject(element as CompileInput));
-    } else if (is.plainObject(value)) {
-      res[field] = getFilteredObject(value);
-    } else if (!is.undefined(value)) {
-      res[field] = value;
+        .map((element) => proxyCompileInput(element as CompileInput));
     }
-  }
-  return res;
+
+    if (is.plainObject(value)) {
+      return proxyCompileInput(value);
+    }
+
+    return value;
+  },
+};
+
+export function proxyCompileInput(input: CompileInput): CompileInput {
```

Comment:
do we really need to make this public?

Reply:
I think filtering should be tested as separate function, because it's quite important